:sparkles: Add preparation of clusterclass

farodin91 opened the pull request:
**What is the purpose of this pull request/Why do we need it?**

**Issue #, if available:**

**Description of changes:**

**Special notes for your reviewer:**

**Checklist:**
- [x] Documentation updated
- [ ] Unit Tests added
- [ ] E2E Tests added
- [x] Includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)

lubedacht commented:
Check out this example from [Cluster API](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/templates/clusterclass-quick-start.yaml)

You seem to be missing a couple of things, e.g. `clusterclass.spec.workers`

Have you been able to create different cluster with different configurations from your ClusterClass template? If yes can you maybe provide some docs on how you did it?

I've checked out your branch, because I have never used ClusterClass before and wanted to test it. I couldn't really figure out how to do it, therefore I dived a bit into the documentation.

https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass#clusterclass-with-patches

If I'm not wrong here, you are mixing up multiple things. We need to work with JSON patches.

```yaml
jsonPatches:
      - op: add
        path: /spec/template....
        valueFrom:
          variable: xxxx
```

Therefore, you cannot have environment variables in the clusterclass template 
**(except for maybe CLUSTER_CLASS_NAME and NAMESPACE, used with `clusterctl generate yaml`)**
e.g.

Instead of:
```yaml
kind: IonosCloudMachineTemplate
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
metadata:
  name: "${CLUSTER_CLASS_NAME}-template"
  namespace: '${NAMESPACE}'
spec:
  template:
    spec:
      numCores: ${IONOSCLOUD_MACHINE_NUM_CORES}
      memoryMB: ${IONOSCLOUD_MACHINE_MEMORY_MB}
      disk:
        image:
          id: ${IONOSCLOUD_MACHINE_IMAGE_ID}
```
You would then need a patch for that. Something like

```yaml
kind: IonosCloudMachineTemplate
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
metadata:
  name: "${CLUSTER_CLASS_NAME}-template"
  namespace: '${NAMESPACE}'
spec:
  template: {}
...
  patches:
  - name: workerMachineType
    definitions:
    - selector:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: IonosCloudMachineTemplate
        matchResources:
          machineDeploymentClass:
            names:
            - XXXX
      jsonPatches:
      - op: add
        path: /spec/template/spec/numCores
        valueFrom:
          variable: numCPUs
     - op: add
        path: /spec/template/spec/memoryMB
        valueFrom:
          variable: memoryMB
     - op: add
       path: /spec/template/spec/disk/image/id
       valueFrom:
         variable: imageID
...

65278 commented on templates/clusterclass-template.yaml:
```diff
@@ -0,0 +1,625 @@
+          controlPlane: true
+    name: kubeVipPodManifest
+  variables:
+  - metadata: {}
```
`metadata: {}` can be removed here. Aids readability.

65278 commented on templates/clusterclass-template.yaml:
```diff
@@ -0,0 +1,625 @@
+        description: FQDN for the control plane.
+        type: string
+  - metadata: {}
+    name: controlPlanePort
```
Could add a default value here

65278 commented on templates/clusterclass-template.yaml:
```diff
@@ -0,0 +1,625 @@
+        path: /spec/template/spec/controlPlaneEndpoint
+        valueFrom:
+          template: |
+            host: '{{ .controlPlaneHost | default .controlPlaneIpAddr }}'
```
Since both are of type string, you could simply collapse these two into one value (a host can be an ip address).

farodin91 replied:
No. The ip is required in kubevip as the host can be an fqdn.

65278 reviewed (commented):
Hi there, nice work.
I haven't tested this, only read it, and all my comments amount to nitpicking, so from my side this is okay, but it's not my place to greenlight this. My only qualification here is also having written a clusterclass.

farodin91 commented:
I've added a fixed some values and added a bash script for basic testing.

Followup PRs can add:

- end to end testing
- quickstart guide for cluster class

lubedacht commented on templates/clusterclass-template.yaml:
```diff
@@ -0,0 +1,613 @@
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
+        kind: IonosCloudMachineTemplate
+        matchResources:
+          machineDeploymentClass:
+            names:
+            worker
```
```suggestion
          machineDeploymentClass:
            names:
            - worker
```

lubedacht commented on hack/test-cluster-class.sh:
```diff
@@ -0,0 +1,12 @@
+#/bin/bash
+
```
Way too complicated. Use the existing tooling for env replacement. We don't want to do that manually.
Write out both files with `clusterctl generate yaml` instead

Add the missing env variables to the project [example envfile](https://github.com/ionos-cloud/cluster-api-provider-ionoscloud/blob/main/envfile.example)

Please rewrite the bash script to something like this:

```bash
source .envfile
export CLUSTER_CLASS_NAME="xyz"
export NAMESPACE="default"

clusterctl generate yaml --from templates/clusterclass-template.yaml > ...
clusterctl generate yaml --from templates/cluster-template-topology > ...

# validate
clusterctl alpha topology plan -f ... -f.. -o ...

# apply
# kubectl apply -f generated-yaml-clusterclass
# kubectl apply -f generated-yaml-cluster-template
```

The idea is to have a script which will help me understand what I have to do. From what I understand currently. This is what I would have to do in order to make things run.

farodin91 replied:
better?

lubedacht commented on templates/cluster-template-topology.yaml:
```diff
@@ -0,0 +1,55 @@
+    controlPlane:
+      replicas: ${CONTROL_PLANE_MACHINE_COUNT}
+    variables:
+    - name: sshKey
+      value: '${IONOSCLOUD_MACHINE_SSH_KEY}'
```
We already have the variable `IONOSCLOUD_MACHINE_SSH_KEYS`  and we would like to be able to provide multiple SSH keys

```suggestion
    - name: sshKeys
      value: '${IONOSCLOUD_MACHINE_SSH_KEYS}'
```

lubedacht commented on templates/cluster-template-topology.yaml:
```diff
@@ -0,0 +1,55 @@
+    - name: location
+      value: '${CONTROL_PLANE_ENDPOINT_LOCATION}'
+    - name: controlPlaneNumCores
+      value: '${IONOSCLOUD_MACHINE_NUM_CORES}'
```
must be int

lubedacht commented on templates/cluster-template-topology.yaml:
```diff
@@ -0,0 +1,55 @@
+    - name: controlPlaneNumCores
+      value: '${IONOSCLOUD_MACHINE_NUM_CORES}'
+    - name: controlPlaneMemoryMB
+      value: '${IONOSCLOUD_MACHINE_MEMORY_MB}'
```
must be int

lubedacht commented on templates/cluster-template-topology.yaml:
```diff
@@ -0,0 +1,55 @@
+    - name: controlPlaneImageID
+      value: '${IONOSCLOUD_MACHINE_IMAGE_ID}'
+    - name: workerNumCores
+      value: '${IONOSCLOUD_MACHINE_NUM_CORES}'
```
must be int

lubedacht commented on templates/cluster-template-topology.yaml:
```diff
@@ -0,0 +1,55 @@
+    - name: workerNumCores
+      value: '${IONOSCLOUD_MACHINE_NUM_CORES}'
+    - name: workerMemoryMB
+      value: '${IONOSCLOUD_MACHINE_MEMORY_MB}'
```
must be int

lubedacht commented on templates/cluster-template-topology.yaml:
```diff
@@ -0,0 +1,55 @@
+    version: '${KUBERNETES_VERSION}'
+    workers:
+      machineDeployments:
+      - class: ${CLUSTER_CLASS_NAME}-worker
```
Your naming for the MachineDeployement `worker` is different in the clusterclass-template. If I try to run the commands it will let me know that a MachineDeployment with the name `xxx-worker` was not found in ClusterClass XY

farodin91 commented:
@lubedacht Everything should be fixed.

lubedacht commented on hack/test-cluster-class.sh:
```diff
@@ -13,14 +13,27 @@ if [[ ! -f "./.envfile" ]]; then
 fi
 source ./.envfile
```
${ROOT_DIR}/.envfile?

farodin91 replied:
fixed

lubedacht commented on hack/test-cluster-class.sh:
```diff
@@ -13,14 +13,27 @@ if [[ ! -f "./.envfile" ]]; then
+
+echo "Installing calico cni"
+echo "make crs-calico"
+echo "kubectl apply -f ./templates/crs/cni/calico.yaml"
```
You have created a variable with the root project directory. Please use it for specifiying paths for input and output files

farodin91 replied:
fixed

lubedacht commented on hack/test-cluster-class.sh:
```diff
@@ -0,0 +1,39 @@
+
+if [[ ! -d "./output" ]]; then
+    echo "create output directory"
+    mkdir "./output"
```
if [[ ! -d "${ROOT_DIR}/output" ]]...
mkdir ${ROOT_DIR}/output

same for if in line 10

farodin91 replied:
fixed

lubedacht commented on hack/test-cluster-class.sh:
```diff
@@ -2,27 +2,27 @@
+clusterctl generate cluster $CLUSTER_NAME -n $NAMESPACE --from "${ROOT_DIR}/templates/cluster-template-topology-calico.yaml" > $GENERATED_CLUSTER_FILE
 
 echo "Validate generated resources"
 clusterctl alpha topology plan -f $GENERATED_CLUSTER_CLASS_FILE -f $GENERATED_CLUSTER_FILE -o output/
```
${ROOT_DIR}/output/

farodin91 replied:
done

jriedel-ionos commented on hack/test-cluster-class.sh:
```diff
@@ -0,0 +1,39 @@
+echo "Setup Cluster API with ClusterResourceSet and ClusterTopology"
+echo "export EXP_CLUSTER_RESOURCE_SET=\"true\""
+echo "export CLUSTER_TOPOLOGY=\"true\""
+echo "clusterclt init --infrastructure=ionos-cloud"
```
```suggestion
echo "clusterctl init --infrastructure=ionos-cloud"
```

farodin91 replied:
updated

jriedel-ionos commented on templates/clusterclass-template.yaml:
```diff
@@ -0,0 +1,611 @@
+            content: |
+              # Restrict key exchange, cipher, and MAC algorithms, as per sshaudit.com
+              # hardening guide.
+              KexAlgorithms curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha256
+              Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
+              MACs hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-128-etm@openssh.com
+              HostKeyAlgorithms ssh-ed25519,ssh-ed25519-cert-v01@openssh.com,sk-ssh-ed25519@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,rsa-sha2-256,rsa-sha2-512,rsa-sha2-256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com
```
https://github.com/ionos-cloud/cluster-api-provider-ionoscloud/pull/140 will soon be merged. Please sync the ssh config from main then.

jriedel-ionos commented on templates/clusterclass-template.yaml:
```diff
@@ -0,0 +1,611 @@
+          content: |
+            # Restrict key exchange, cipher, and MAC algorithms, as per sshaudit.com
+            # hardening guide.
+            KexAlgorithms curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha256
+            Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
+            MACs hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-128-etm@openssh.com
+            HostKeyAlgorithms ssh-ed25519,ssh-ed25519-cert-v01@openssh.com,sk-ssh-ed25519@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,rsa-sha2-256,rsa-sha2-512,rsa-sha2-256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com
```
^ same as above

farodin91 replied:
updated
